src/data/sentencepiece_vocab.cpp

```diff
@@ -221,6 +217,24 @@ class SentencePieceVocab : public IVocab {
       words.push_back(getEosId());
     return words;
   }
+  
+  std::vector<SourceToken> encodePreservingSource(const std::string& line, bool addEOS, bool inference) const override {
```

Comment:
We need Words to come out.  Don't roll it into the same vector.  Just give alignment as side information.